pkg/util/deployment/deployment.go

```diff
@@ -444,3 +444,35 @@ func WaitForObservedDeployment(getDeploymentFunc func() (*extensions.Deployment,
 		return deployment.Status.ObservedGeneration >= desiredGeneration, nil
 	})
 }
+
+// ResolveFenceposts resolves both maxSurge and maxUnavailable. This needs to happen in one
+// step. For example:
+//
+// 2 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1), then old(-1), then new(+1)
+// 1 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1)
+// 2 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
+// 1 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1)
+// 2 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
+// 1 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1)
+func ResolveFenceposts(maxSurge, maxUnavailable *intstrutil.IntOrString, desired int) (int, int, error) {
+	surge, err := intstrutil.GetValueFromIntOrPercent(maxSurge, desired, true)
+	if err != nil {
+		return 0, 0, err
+	}
+	if surge == 0 {
+		// If a user has specified maxSurge to zero but maxUnavailable as a non-zero percent,
+		// then we want to allow scaling down and not block rollouts.
```

Comment:
Sorry, was reading on the API fields, not the code.

On Fri, Mar 4, 2016 at 12:37 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/util/deployment/deployment.go
> <https://github.com/kubernetes/kubernetes/pull/22516#discussion_r55069493>
> :
>
> > +// step. For example:
> > +//
> > +// 2 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1), then old(-1), then new(+1)
> > +// 1 desired, max unavailable 1%, surge 0% - should scale old(-1), then new(+1)
> > +// 2 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
> > +// 1 desired, max unavailable 25%, surge 1% - should scale new(+1), then old(-1)
> > +// 2 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1), then new(+1), then old(-1)
> > +// 1 desired, max unavailable 0%, surge 1% - should scale new(+1), then old(-1)
> > +func ResolveFenceposts(maxSurge, maxUnavailable *intstrutil.IntOrString, desired int) (int, int, error) {
> > +	surge, err := intstrutil.GetValueFromIntOrPercent(maxSurge, desired, true)
> > +	if err != nil {
> > +		return 0, 0, err
> > +	}
> > +	if surge == 0 {
> > +		// If a user has specified maxSurge to zero but maxUnavailable as a non-zero percent,
> > +		// then we want to allow scaling down and not block rollouts.
>
> Sorry for not being clear.
>
> The current code rounds maxUnavailable down and maxSurge up. I was
> proposing to leave them that way.
>
> But if both unavailable pods and surge pods end up being zero, set
> unavailable to 1, on the theory that surge might not work due to quota.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22516/files#r55069493>.
>